### PR TITLE
Allow #[test_proc] to expect specific assertion failures

### LINKED ISSUE #1071: [DSLX] Feature: Add `#[test_fail]` and `#[test_proc_fail]`
Right now `#[test_proc]` and `#[test]` always has to succeed.

Adding new language testing feature will allow for negative testing.
I think it would only require adding  `#[test_fail]` and `#[test_proc_fail]` to the language parser.
Test procs or fn should keep their current type format, and the bulk of the changes would be
applied to the DSLX interpreter.

### PR DESCRIPTION
**Pull request by rw1nkler**
This commit introduces the ability to specify that a test proc is expected to fail on a particular `assert!` or `fail!` in the DSLX interpreter. The `#[test_proc]` attribute accepts an optional `expected_fail` parameter. with the label marking the failure. If a failure occurs with the matching label, the test is treated as successful.

The choice to implement this as an attribute, rather than a built-in, as discussed in https://github.com/google/xls/issues/1071#issuecomment-1642453871 was motivated by cases where there's no clear location to “catch” the failure. For example, a test that spawns a proc containing only a single `fail!` statement.

The functionality is implemented by storing the `expected_fail` value in the `TestProc` object. When an `assert!` or `fail!` is handled in the DSLX interpreter, the label of the failure is attached to the error using an `absl::Status` payload. This is then compared in `RunDslxTestProc` against the stored label in the `TestProc`.

Resolves #1071

### CODE REVIEW

File: docs_src/tutorials/how_to_use_procs.md [hunk 1]
```diff
@@ -217,6 +217,11 @@ test proc is very similar to a normal proc with the following changes:
     input channel for terminating interpretation. When the test is complete, the
     proc should send the test's status (`true` on success, `false` on failure)
     on that channel (commonly called the "terminator" channel).
+*   The `#[test_proc]` attribute also supports an optional `expected_fail`
+    parameter. This allows writing tests that are expected to fail due to
+    a specific `assert!` or `fail!` condition. The value must match the label
+    on the failing assertion. If a failure with the matching label occurs,
+    the test is treated as successful.
```
Comment (proppy) [inline docs_src/tutorials/how_to_use_procs.md:224]: "and if no failure happens: the test is treated as failed" (or add "if and only if")

**Comment by proppy** [comment]
curious if that's something that regular SystemVerilog/UVM support, and what this would ultimately map to if we were generating SV test benches from DSLX tests.

**Comment by rw1nkler** [comment]
> curious if that's something that regular SystemVerilog/UVM support, and what this would ultimately map to if we were generating SV test benches from DSLX tests.

I believe there is no built-in option to ignore the `$fatal` statement in SystemVerilog. However, if support for `fail!` and `assert!` is extended to allow the generation of a signal upon failure (as proposed in [google/xls#1352](https://github.com/google/xls/issues/1352)), the testbench could include logic to respond accordingly by either triggering a failure or allowing the test to pass based on the context.

**Comment by meheff** [comment]
@richmckeever mind having a look at this as it touches on the frontend. The feature is very nice to have.

File: docs_src/tutorials/how_to_use_procs.md (hunk 1 above)
Comment (rw1nkler) [reply docs_src/tutorials/how_to_use_procs.md]: fixed

File: xls/dslx/frontend/parser.cc [hunk 2]
```diff
@@ -815,8 +815,30 @@ Parser::ParseAttribute(absl::flat_hash_map<std::string, Function*>* name_to_fn,
...
+    if (peek->kind() == TokenKind::kOParen) {
+      XLS_RETURN_IF_ERROR(DropTokenOrError(TokenKind::kOParen));
+      XLS_ASSIGN_OR_RETURN(Token parameter_name,
+                           PopTokenOrError(TokenKind::kIdentifier));
+
+      if (parameter_name.GetStringValue() == "expected_fail") {
+        XLS_RETURN_IF_ERROR(DropTokenOrError(TokenKind::kEquals));
+        XLS_ASSIGN_OR_RETURN(Token parameter_value,
+                             PopTokenOrError(TokenKind::kString));
+        expected_fail =
```
Comment (richmckeever) [inline xls/dslx/frontend/parser.cc:829]: I think you can just assign this without explicitly constructing the optional.

File: xls/dslx/frontend/parser.cc [hunk 3]
```diff
@@ -815,8 +815,30 @@ Parser::ParseAttribute(absl::flat_hash_map<std::string, Function*>* name_to_fn,
         peek->span(), absl::StrCat("Invalid test type: ", peek->ToString()));
   }
   if (attribute_name == "test_proc") {
+    std::optional<std::string> expected_fail = std::nullopt;
```
Comment (richmckeever) [inline xls/dslx/frontend/parser.cc:818]: This should probably be a std::optional<bool> and we convert the string (if present) to a bool in here.

File: xls/dslx/frontend/proc.cc [hunk 4]
```diff
@@ -154,6 +154,10 @@ std::string ProcLike::ToString() const {
 TestProc::~TestProc() = default;
 
 std::string TestProc::ToString() const {
+  if (expected_fail_.has_value()) {
+    return absl::StrFormat("#[test_proc(expected_fail=\"%s\")]\n%s",
+                           expected_fail_.value(), proc_->ToString());
```
Comment (richmckeever) [inline xls/dslx/frontend/proc.cc:159]: Prefer `*expected_fail_` rather than `expected_fail.value()` generally when accessing optionals.

File: xls/dslx/frontend/proc.h [hunk 5]
```diff
@@ -203,7 +203,9 @@ class TestProc : public AstNode {
  public:
   static std::string_view GetDebugTypeName() { return "test proc"; }
 
-  TestProc(Module* owner, Proc* proc) : AstNode(owner), proc_(proc) {}
+  TestProc(Module* owner, Proc* proc,
+           std::optional<std::string> expected_fail = std::nullopt)
```
Comment (richmckeever) [inline xls/dslx/frontend/proc.h:207]: Let's take this as a std::optional<bool> here and store it that way; propagating the string seems more error prone. If someday it's more than a boolean, we can use an enum or struct at that point.

File: xls/dslx/run_routines/run_routines.cc [hunk 6]
```diff
@@ -202,9 +202,24 @@ absl::Status RunDslxTestProc(ImportData* import_data, TypeInfo* type_info,
       hierarchy_interpreter->GetInterfaceChannel(0);
 
   // Run until a single output appears in the terminal channel.
-  XLS_RETURN_IF_ERROR(
+  absl::Status status =
       hierarchy_interpreter->TickUntilOutput({{terminal_channel_name, 1}})
-          .status());
+          .status();
+  auto label_opt = GetAssertionLabelFromError(status);
+  if (status != absl::OkStatus() && tp->expected_fail().has_value() &&
```
Comment (richmckeever) [inline xls/dslx/run_routines/run_routines.cc:209]: Prefer `!status.ok()` instead of comparing to OkStatus().

File: xls/dslx/run_routines/run_routines.cc [hunk 7]
```diff
@@ -202,9 +202,24 @@ absl::Status RunDslxTestProc(ImportData* import_data, TypeInfo* type_info,
...
+      label_opt.has_value()) {
+    std::string label = label_opt.value();
```
Comment (richmckeever) [inline xls/dslx/run_routines/run_routines.cc:211]: Here and below, use * instead of .value().

File: xls/dslx/errors.cc [hunk 8]
```diff
@@ -33,6 +33,8 @@
 namespace xls::dslx {
 namespace {
 
+static constexpr std::string_view label_payload_name = "label";
```
Comment (richmckeever) [inline xls/dslx/errors.cc:36]: The conventional name here would be kLabelPayloadName. Also drop `static` because being in an anonymous namespace has the same effect.

File: xls/dslx/errors.cc [hunk 9]
```diff
@@ -63,6 +65,25 @@ absl::Status FailureErrorStatus(const Span& span, std::string_view message,
...
+                                            std::string_view message,
+                                            const FileTable& file_table) {
+  absl::Status error = FailureErrorStatus(span, message, file_table);
+  error.SetPayload(label_payload_name, absl::Cord(label));
+  return error;
+}
+
+std::optional<std::string> GetAssertionLabelFromError(
+    const absl::Status& status) {
+  auto label_opt = status.GetPayload(label_payload_name);
```
Comment (richmckeever) [inline xls/dslx/errors.cc:79]: Prefer writing the actual type instead of `auto`, except for iterators, destructured tuples, and objects that don't have a practically expressible type.

File: xls/dslx/run_routines/run_routines_test.cc [hunk 10]
```diff
@@ -533,6 +533,37 @@ proc doomed {
               IsTestResult(TestResult::kParseOrTypecheckError, 0, 0, 0));
 }
 
+TEST_P(RunRoutinesTest, TestProcWithExpectedFail) {
```
Comment (richmckeever) [inline xls/dslx/run_routines/run_routines_test.cc:536]: Let's test this with both fail! and assert! in 2 separate tests (this one is really the assert! one).

**Review by richmckeever — CHANGES_REQUESTED** [review]

File: xls/dslx/frontend/parser.cc (hunk 3 above)
Comment (rw1nkler) [reply xls/dslx/frontend/parser.cc:818]: I'm not sure if I understand you correctly.

The string is used later when comparing the label of the expected failure with the actual label from the assertion. This way we can still fail on assertions different then the expected one. Substituting it with a bool can cause false positives in tests.

The comparison takes place in: 
https://github.com/google/xls/pull/2109/files#diff-4b75394b54655a9c1d63cc7c6e63bee976945b37697c3a268df6a51e5dae59b6R213

File: xls/dslx/frontend/proc.h (hunk 5 above)
Comment (rw1nkler) [reply xls/dslx/frontend/proc.h:207]: The explanation for putting here a string instead of bool is described in https://github.com/google/xls/pull/2109/files#r2088996980

File: xls/dslx/frontend/parser.cc (hunk 3 above)
Comment (richmckeever) [reply xls/dslx/frontend/parser.cc:818]: Yeah, I think I was thrown off by the variable name. It might be good to name it "expected_fail_label" both here and in the AST.

Comment (rw1nkler) [reply xls/dslx/frontend/parser.cc]: I changed the name to ` expected_fail_label` as it was suggested

File: xls/dslx/frontend/proc.cc (hunk 4 above)
Comment (rw1nkler) [reply xls/dslx/frontend/proc.cc]: fixed

File: xls/dslx/run_routines/run_routines.cc (hunk 6 above)
Comment (rw1nkler) [reply xls/dslx/run_routines/run_routines.cc]: fixed

File: xls/dslx/run_routines/run_routines.cc (hunk 7 above)
Comment (rw1nkler) [reply xls/dslx/run_routines/run_routines.cc]: fixed

File: xls/dslx/errors.cc (hunk 9 above)
Comment (rw1nkler) [reply xls/dslx/errors.cc]: fixed

File: xls/dslx/run_routines/run_routines_test.cc (hunk 10 above)
Comment (rw1nkler) [reply xls/dslx/run_routines/run_routines_test.cc]: I added a separate test for the `fail!` macro

**Comment by rw1nkler** [comment]
I also modified the DSLX formater accordingly

File: xls/dslx/run_routines/run_routines_test.cc [hunk 11]
```diff
@@ -533,6 +533,68 @@ proc doomed {
               IsTestResult(TestResult::kParseOrTypecheckError, 0, 0, 0));
 }
 
+TEST_P(RunRoutinesTest, TestProcExpectedToFailOnAssert) {
```
Comment (richmckeever) [inline xls/dslx/run_routines/run_routines_test.cc:536]: Let's add one more test, where we assert with an unexpected label and the test result is a failure.

Comment (rw1nkler) [reply xls/dslx/run_routines/run_routines_test.cc:536]: I added separate tests for both `assert!` and `fail!` macros that check whether the failure on incorrect label is treated correctly.

**Review by richmckeever — APPROVED** [review]

File: xls/dslx/errors.cc [hunk 12]
```diff
@@ -63,6 +65,25 @@ absl::Status FailureErrorStatus(const Span& span, std::string_view message,
...
+                                            std::string_view message,
+                                            const FileTable& file_table) {
+  absl::Status error = FailureErrorStatus(span, message, file_table);
+  error.SetPayload(kLabelPayloadName, absl::Cord(label));
+  return error;
+}
+
+std::optional<std::string> GetAssertionLabelFromError(
+    const absl::Status& status) {
+  absl::optional<absl::Cord> label_opt = status.GetPayload(kLabelPayloadName);
```
Comment (proppy) [inline xls/dslx/errors.cc:79]: std::optional?

Comment (rw1nkler) [reply xls/dslx/errors.cc]: fixed, but we cannot return `status.GetPayload(kLabelPayloadName) directly`

File: xls/dslx/frontend/parser.h [hunk 13]
```diff
@@ -619,7 +619,9 @@ class Parser : public TokenParser {
   absl::StatusOr<TestFunction*> ParseTestFunction(Bindings& bindings,
                                                   const Span& attribute_span);
 
-  absl::StatusOr<TestProc*> ParseTestProc(Bindings& bindings);
+  absl::StatusOr<TestProc*> ParseTestProc(
+      Bindings& bindings,
+      std::optional<std::string> expected_fail_label = std::nullopt);
```
Comment (dplassgit) [inline xls/dslx/frontend/parser.h:624]: Optional: ` = std::nullopt` isn't needed, since all the uses already specify the fail label.

**Comment by rw1nkler** [comment]
Please let me know if I need to update anything else

**Review by dplassgit — APPROVED** [review]

**Review by proppy — APPROVED** [review]

**Comment by proppy** [comment]
should we also add support for `excepted_fail` for regular function `#[test]`?
